### PR TITLE
ci: fix codecov upload issue

### PR DESCRIPTION
**Pull request by shahar-h**
**What this PR does / why we need it**:
Fixed codecov upload issue by configuring authentication with [oidc](https://github.com/codecov/codecov-action?tab=readme-ov-file#using-oidc).
In case PR is from fork OIDC will not work since PRs from forks have `read` permission on `id-token`.
Anyway Codecov supports [Token-less Uploads](https://docs.codecov.com/docs/codecov-uploader#supporting-token-less-uploads-for-forks-of-open-source-repos-using-codecov) for PRs from forks.
Also set `fail_ci_if_error` back to `true`.

```log
['error'] There was an error running the uploader: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 429 - {'detail': ErrorDetail(string='Rate limit reached. Please upload with the Codecov repository upload token to resolve issue. Expected time to availability: 2032s.', code='throttled')}
```

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

**Comment by shahar-h** [comment]
Works as expected: 
```log
info - 2024-05-17 15:18:33,306 -- The PR is happening in a forked repo. Using tokenless upload.
```

File: .github/workflows/build_and_test.yaml [hunk 1]
```diff
@@ -52,12 +55,16 @@ jobs:
...
       with:
-        fail_ci_if_error: false
+        fail_ci_if_error: true
         files: ./coverage.xml
         name: codecov-envoy-gateway
         verbose: true
+        # only use oidc for push events or PRs from the same repo, since PRs from forks cannot fetch an OIDC token
+        # Codecov supports Token-less Uploads for PRs from forks:
+        # https://docs.codecov.com/docs/codecov-uploader#supporting-token-less-uploads-for-forks-of-open-source-repos-using-codecov
+        use_oidc: ${{ github.event_name == 'push' || (github.event_name == 'pull_request' && github.event.pull_request.head.repo.full_name == github.repository) }}
```
Comment (arkodg) [inline .github/workflows/build_and_test.yaml:67]: should we keep this simple - event is push ? so we encourage PRs via forks only

Comment (shahar-h) [reply .github/workflows/build_and_test.yaml]: updated

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Comment by shahar-h** [comment]
According to https://github.com/codecov/codecov-action/pull/1404 `use_oidc` should be ignored on pull request from forks, and instead tokenless request should be sent. That's what I did in my previous [PR](https://github.com/envoyproxy/gateway/pull/3419) and it failed: https://github.com/envoyproxy/gateway/actions/runs/9129469887/job/25104103828?pr=3419.
Opened an [issue](https://github.com/codecov/codecov-action/issues/1435).

**Comment by shahar-h** [comment]
@arkodg we can switch to `use_oidc: true` once https://github.com/codecov/codecov-action/pull/1437 is merged and a new codecov version is released.

**Comment by shahar-h** [comment]
@arkodg this won't work for `dependabot` PRs since they are not from fork, so I suggest to go back to the previous condition: 
```yaml
use_oidc: ${{ github.event_name == 'push' || (github.event_name == 'pull_request' && github.event.pull_request.head.repo.full_name == github.repository) }}`
```
Once `codecov-action` [fix](https://github.com/envoyproxy/gateway/pull/3421#issuecomment-2118327791) will be merged we can change that to true.

**Review by shawnh2 — APPROVED** [review]
lgtm, thanks for the fix

**Comment by shahar-h** [comment]
@zirain @shawnh2 this won't work for dependabot PRs as I mentioned [here](https://github.com/envoyproxy/gateway/pull/3421?notification_referrer_id=NT_kwDOAPydGbQxMDcwNjUwOTEwNDoxNjU1NTI4OQ#issuecomment-2119107790). I'll open a new PR to fix that.

**Comment by shahar-h** [comment]
See https://github.com/envoyproxy/gateway/pull/3439

**Pull request merged**